src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/MergeDataToSearchResultOperation.java

```diff
@@ -34,36 +44,111 @@ public SearchResultView execute(final CaseType caseType,
+    private List<SearchResultViewColumn> prepareSearchResultViewColumns(CaseType caseType,
+                                                                        SearchResult searchResult) {
+        return Arrays.stream(searchResult.getFields())
+            .flatMap(searchResultField -> caseType.getCaseFields().stream()
+                    .filter(caseField -> caseField.getId().equals(searchResultField.getCaseFieldId()))
+                    .map(caseField -> new SearchResultViewColumn(
+                        buildCaseFieldId(searchResultField),
+                        getNestedCaseField(caseField, searchResultField).getFieldType(),
+                        searchResultField.getLabel(),
+                        searchResultField.getDisplayOrder(),
+                        searchResultField.isMetadata()))
+                    )
+            .collect(Collectors.toList());
+    }
+
+    private String buildCaseFieldId(SearchResultField searchResultField) {
+        if (StringUtils.isNotBlank(searchResultField.getCaseFieldPath())) {
+            return searchResultField.getCaseFieldId() + '.' + searchResultField.getCaseFieldPath();
+        } else {
+            return searchResultField.getCaseFieldId();
+        }
+    }
+
+    private CaseField getNestedCaseField(final CaseField caseField,
+                                         final SearchResultField searchResultField) {
+        return caseField.findNestedElementByPath(searchResultField.getCaseFieldPath());
+    }
+
+    private SearchResultViewItem buildSearchResultViewItem(final CaseDetails caseDetails,
+                                                           final CaseType caseType,
+                                                           final SearchResult searchResult) {
+
+        String caseId = caseDetails.hasCaseReference() ? caseDetails.getReferenceAsString() : caseDetails.getId();
+
+        Map<String, JsonNode> caseData = new HashMap<>(caseDetails.getData());
+        Map<String, Object> caseMetadata = new HashMap<>(caseDetails.getMetadata());
+        Map<String, TextNode> labels = getLabelsFromCaseFields(caseType);
+
+        Map<String, Object> caseFields = prepareData(searchResult, caseData, caseMetadata, labels);
+
+        return new SearchResultViewItem(caseId, caseFields);
+    }
+
+    private Map<String, Object> prepareData(SearchResult searchResult,
+                                            Map<String, JsonNode> caseData,
+                                            Map<String, Object> metadata,
+                                            Map<String, TextNode> labels) {
+
+        Map<String, Object> newResults = new HashMap<>();
+
+        List<SearchResultField> searchResultsWithPath = Arrays.stream(searchResult.getFields())
+            .filter(e -> StringUtils.isNotBlank(e.getCaseFieldPath()))
+            .collect(Collectors.toList());
+
+        searchResultsWithPath.forEach(searchResultField -> {
+            caseData.forEach((key, value) -> {
```

Comment:
rather than a foreach, can't we just do:
            `JsonNode jsonNode = caseData.get(searchResultField.getCaseFieldId());`
?